File: src/conversation_graph.py
from typing import Dict, TypedDict, List, Annotated, Union
from langgraph.graph import Graph, START
from langchain_anthropic import ChatAnthropic
from langchain_core.messages import HumanMessage, AIMessage
from langchain.memory import ConversationBufferMemory
import base64

# Define our state structure
class ConversationState(TypedDict):
    messages: List[Union[HumanMessage, AIMessage]]
    memory: ConversationBufferMemory
    image_analysis: Dict[str, str]

def create_conversation_graph(api_key: str):
    # Initialize Claude with vision capabilities
    model = ChatAnthropic(
        model="claude-3-sonnet-20240229",
        anthropic_api_key=api_key,
        max_tokens=4096
    )
    
    # Define our nodes
    def analyze_images(state: ConversationState) -> Dict:
        print("Debug: Entering analyze_images")
        
        latest_message = state["messages"][-1]
        if hasattr(latest_message, "additional_kwargs") and "image_data" in latest_message.additional_kwargs:
            try:
                # Convert image data to base64
                image_data = latest_message.additional_kwargs["image_data"]
                image_base64 = base64.b64encode(image_data).decode('utf-8')
                
                # Create message with properly formatted image
                image_message = HumanMessage(
                    content=[
                        {
                            "type": "text",
                            "text": "What do you see in this image?"
                        },
                        {
                            "type": "image",
                            "source": {
                                "type": "base64",
                                "media_type": "image/jpeg",
                                "data": image_base64
                            }
                        }
                    ]
                )
                
                # Get Claude's analysis
                response = model.invoke([image_message])
                analysis = response.content
                state["image_analysis"][str(len(state["image_analysis"]))] = analysis
                
                # Add the analysis as a response
                state["messages"].append(AIMessage(content=analysis))
                print(f"Debug: Image analysis complete: {analysis}")
            except Exception as e:
                print(f"Debug: Error in image analysis: {str(e)}")
                raise e
                
        return state

    def generate_response(state: Dict) -> Dict:
        print("Debug: Entering generate_response")
        
        try:
            # Only generate a new response if the last message wasn't from image analysis
            if not (state["messages"] and isinstance(state["messages"][-1], AIMessage)):
                context = ""
                if state["image_analysis"]:
                    context = "Previous image analysis:\n" + "\n".join(state["image_analysis"].values()) + "\n\n"
                
                latest_message = state["messages"][-1]
                prompt = context + latest_message.content
                
                response = model.invoke([HumanMessage(content=prompt)])
                state["messages"].append(AIMessage(content=response.content))
            
            return state
            
        except Exception as e:
            print(f"Debug: Error in generate_response: {str(e)}")
            raise e

    # Create and compile the graph
    workflow = Graph()
    
    workflow.add_node("analyze_images", analyze_images)
    workflow.add_node("generate_response", generate_response)
    
    # Add edges - only use add_edge() for START
    workflow.add_edge(START, "analyze_images")
    workflow.add_edge("analyze_images", "generate_response")
    
    # Set final node
    workflow.set_finish_point("generate_response")
    
    return workflow.compile() 

File: typescript/src/conversationGraph.ts
import { Anthropic } from "@anthropic-ai/sdk";
import type { Message } from '@anthropic-ai/sdk/resources/messages';
import { ConversationState, ContentBlock, StoredMessage, MediaType } from "./types";
import * as fs from 'fs';
import { promisify } from 'util';

const readFile = promisify(fs.readFile);

export class ConversationGraph {
    private model: Anthropic;
    private state: ConversationState;

    constructor(apiKey: string) {
        this.model = new Anthropic({
            apiKey: apiKey
        });

        this.state = {
            messages: [],
            imageAnalysis: {}
        };
    }

    private async analyzeImage(imageData: Buffer): Promise<string> {
        try {
            const base64Image = imageData.toString('base64');
            
            // Debug logging
            console.log("Image size:", imageData.length, "bytes");
            console.log("MIME type:", this.getMimeType(imageData));
            
            const content: ContentBlock[] = [
                {
                    type: "text",
                    text: "Please provide a detailed analysis of this image, including:\n" +
                          "1. All visible section headings\n" +
                          "2. Key rules and mechanics mentioned\n" +
                          "3. Any specific numerical values or modifiers\n" +
                          "4. Tables or structured data\n" +
                          "5. Important terms and definitions\n" +
                          "Please be as specific as possible in extracting and organizing this information."
                },
                {
                    type: "image",
                    source: {
                        type: "base64",
                        media_type: this.getMimeType(imageData),
                        data: base64Image
                    }
                }
            ];

            const response = await this.model.messages.create({
                model: "claude-3-sonnet-20240229",
                max_tokens: 4096,
                messages: [{
                    role: "user",
                    content
                }]
            });

            return response.content[0].text;
        } catch (error) {
            console.error("Image analysis error details:", error);
            throw error;
        }
    }

    // Helper method to detect MIME type
    private getMimeType(buffer: Buffer): MediaType {
        const signatures: Record<string, MediaType> = {
            'ffd8ffe0': 'image/jpeg',
            '89504e47': 'image/png',
            '47494638': 'image/gif',
            '52494646': 'image/webp'
        };

        // Get first 4 bytes as hex
        const hex = buffer.toString('hex', 0, 4);
        
        // Log for debugging
        console.log("File signature:", hex);
        
        for (const [signature, mimeType] of Object.entries(signatures)) {
            if (hex.startsWith(signature)) {
                return mimeType;
            }
        }
        
        return 'image/jpeg'; // default fallback
    }

    private async generateResponse(userMessage: string): Promise<string> {
        try {
            const context = Object.values(this.state.imageAnalysis).length > 0
                ? "Previous image analysis:\n" + Object.values(this.state.imageAnalysis).join("\n\n") + "\n\n"
                : "";

            const response = await this.model.messages.create({
                model: "claude-3-sonnet-20240229",
                max_tokens: 4096,
                messages: [{
                    role: "user",
                    content: userMessage || "Hello"  // Provide default message if empty
                }]
            });

            return response.content[0].text;
        } catch (error) {
            console.error("Error generating response:", error);
            throw error;
        }
    }

    async processMessage(input: string | Buffer): Promise<string> {
        try {
            let response: string;

            if (Buffer.isBuffer(input)) {
                response = await this.analyzeImage(input);
                this.state.imageAnalysis[String(Object.keys(this.state.imageAnalysis).length)] = response;
            } else {
                response = await this.generateResponse(input);
            }

            // Store the interaction using our StoredMessage type
            const userMessage: StoredMessage = {
                role: "user",
                content: typeof input === 'string' ? input : "Uploaded an image"
            };

            const assistantMessage: StoredMessage = {
                role: "assistant",
                content: response
            };

            this.state.messages.push(userMessage, assistantMessage);

            return response;

        } catch (error) {
            console.error('Error processing message:', error);
            throw error;
        }
    }

    getState(): ConversationState {
        return this.state;
    }
}
